### ISSUE TITLE
Send the download/upload updates with a number of milliseconds apart

### ISSUE THREAD

**Issue by freethinker6** [akinaru/speed-test-lib#3]
Hi, thank you for creating this library, works great! I have one suggestion, could you please add a new mode where the time duration for the test is settable for small files? The test will have to be restarted after every cycle finishes so I can update the user after each cycle. This will be nice for UI animations and such. I am trying to test this against the mobile network so the users don't spend too much from their data plan.
Regards.

**Comment by akinaru** [comment]
Hello,

I've released v1.08 with the requested functionality. Let me know if this help 

The following will start a download on the specified host/uri/port, and will restart the download when it is finished until `repeatWindow` is elapsed. `reportPeriodMillis` is the number of millis between each report dispatched in `repeatListener`, an interface that will also tell you when your time window is finished : 

``` 
    /**
     * Start repeat download task
     *
     * @param hostname           server hostname
     * @param port               server port
     * @param uri                uri to fetch to download file
     * @param repeatWindow       time window for the repeated download in milliseconds
     * @param reportPeriodMillis time interval between each report in milliseconds
     * @param repeatListener     listener for download repeat task completion & reports
     */
    public void startDownloadRepeat(String hostname,
                                    int port,
                                    String uri,
                                    int repeatWindow,
                                    int reportPeriodMillis,
                                    IRepeatListener repeatListener)
```
I need an extra Thread to force close socket in the main ThreadPool. That will give you 2 Thread used by Speed Test library instead of one initially if you use this method; but I added an additionnal report callback so you dont have to create a timertask yourself.

You can use it as following (window = 11s, report interval = 1s)

```
        speedTestSocket.startDownloadRepeat("1.testdebit.info", 80, "/fichiers/10Mo.dat", 11000, 1000, new IRepeatListener() {
            @Override
            public void onFinish(SpeedTestReport report) {
                System.out.println("--------------- FINISH REPORT -----------------------------");
                System.out.println("progress             : " + report.getProgressPercent() + "%");
                System.out.println("transfer rate bit    : " + report.getTransferRateBit() + "b/s");
                System.out.println("transfer rate octet  : " + report.getTransferRateOctet() + "B/s");
                System.out.println("uploaded for now     : " + report.getTemporaryPacketSize() + "/" + report.getTotalPacketSize());
                System.out.println("amount of time       : " + ((report.getReportTime() - report.getStartTime()) / 1000) + "s");
                System.out.println("request number       : " + speedTestSocket.getLiveDownloadReport().getRequestNum());
                System.out.println("--------------------------------------------------------");
            }

            @Override
            public void onReport(SpeedTestReport report) {
                System.out.println("--------------- report -----------------------------");
                System.out.println("progress             : " + report.getProgressPercent() + "%");
                System.out.println("transfer rate bit    : " + report.getTransferRateBit() + "b/s");
                System.out.println("transfer rate octet  : " + report.getTransferRateOctet() + "B/s");
                System.out.println("uploaded for now     : " + report.getTemporaryPacketSize() + "/" + report.getTotalPacketSize());
                System.out.println("amount of time       : " + ((report.getReportTime() - report.getStartTime()) / 1000) + "s");
                System.out.println("request number       : " + speedTestSocket.getLiveDownloadReport().getRequestNum());
                System.out.println("--------------------------------------------------------");
            }
        });
```
I adapted the report for this mode so you will have the total packet received for all requests, number of requests, percentage based on time window.

This can be useful if you need to target a server which has API for small files only like you said.

I added another api to force stop download  & upload task : 

```
speedTestSocket.forceStopTask();
```

The following code will download a 100Mo file and stop the download after 5 seconds & request reports with an interval of 400ms:

```
final Timer timer = new Timer();

TimerTask reportTask = new TimerTask() {
    @Override
    public void run() {

        SpeedTestReport downloadReport = speedTestSocket.getLiveDownloadReport();
        System.out.println("---------------current download report--------------------");
        System.out.println("progress             : " + downloadReport.getProgressPercent() + "%");
        System.out.println("transfer rate bit    : " + downloadReport.getTransferRateBit() + "b/s");
        System.out.println("transfer rate octet  : " + downloadReport.getTransferRateOctet() + "B/s");
        System.out.println("downloaded for now   : " + downloadReport.getTemporaryPacketSize() + "/" + downloadReport.getTotalPacketSize());
        System.out.println("amount of time       : " + ((downloadReport.getReportTime() - downloadReport.getStartTime()) / 1000) + "s");
    }
};

// scheduling the task at interval
timer.scheduleAtFixedRate(reportTask, 0, 400);

TimerTask stopTask = new TimerTask() {
    @Override
    public void run() {
        System.out.println("--------------- FINISH REPORT -----------------------------");
        SpeedTestReport downloadReport = speedTestSocket.getLiveDownloadReport();
        System.out.println("---------------current download report--------------------");
        System.out.println("progress             : " + downloadReport.getProgressPercent() + "%");
        System.out.println("transfer rate bit    : " + downloadReport.getTransferRateBit() + "b/s");
        System.out.println("transfer rate octet  : " + downloadReport.getTransferRateOctet() + "B/s");
        System.out.println("downloaded for now   : " + downloadReport.getTemporaryPacketSize() + "/" + downloadReport.getTotalPacketSize());
        System.out.println("amount of time       : " + ((downloadReport.getReportTime() - downloadReport.getStartTime()) / 1000) + "s");

        speedTestSocket.forceStopTask();
        
        if (timer != null) {
            timer.cancel();
            timer.purge();
        }
    }
};
timer.schedule(stopTask, 5000);

speedTestSocket.startDownload("1.testdebit.info", 80, "/fichiers/100Mo.dat");
```

`forceStopTask` will force download/upload task to stop which will close the socket and may trigger the `onDownloadError` or `onUploadError` for a `socket error` message but nothing to worry about. 

Using this second method or the first one to download during a fix time will however be the same for the user data plan.

For the upload method, doing the same as download repeat is not needed because content-length is set by the client (whereas for download you may not have an uri that match your need for small/big files)

You can test those features with : `compile 'com.github.akinaru:speedtest:1.08'`

**Comment by freethinker6** [comment]
Hello,



Thanks for your quick reply.



I think the 2nd approach will work best for my case. The only things I can
report back to you so far are:



1.       In the first approach, with the startDownloadRepeat method I get
zero for the transfer rates:

progress             : 81.82727%

transfer rate bit    : 0.0b/s

transfer rate octet  : 0.0B/s

uploaded for now     : 6294256.0/10000000

amount of time       : 9s

request number       : 0



2.       As for the second approach, with the stopTask, I am getting the
first notification as follows:

progress             : 0.0%

transfer rate bit    : 0.0b/s

transfer rate octet  : 0.0B/s

downloaded for now   : 0.0/0

amount of time       : 1463149679s



3.       Would it be possible to add a new SpeedTestError type to know if
the error is coming from stopping the download/upload task? I am getting
CONNECTION_ERROR with the message “Socket Closed”. Also, when I stop an
update task, I receive two Socket errors instead of just one from stopping
the task.

I will appreciate your comments. Thanks.


Regards,

Paul

2016-05-12 20:44 GMT-05:00 Bertrand Martel <notifications@github.com>:

**Comment by akinaru** [comment]
Thank you for your feedback, I've fixed most of these errors in v1.09 : 
* zero for the transfer rates : I wrote bug fix + set correct report values for start time when download has not begun yet
* amount of time incorrect when download has not begun : if one of your first notifications arrived and download has not begun yet. start time will be 0. So amount of time cant be calculated, this is why you have some strange value here. I corrected it in the example with:
```
if (downloadReport.getStartTime() > 0) {
    System.out.println("amount of time       : "
            + ((downloadReport.getReportTime() - downloadReport.getStartTime()) / 1000) + "s");
}
```
If `startTime` is 0, you can be sure download has not begun

* I created a speed test error in case of ` speedTestSocket.forceStopTask();` : `SpeedTestError.FORCE_CLOSE_SOCKET`

If you encounter an error other than the latter, you have a true socket error :
```
@Override
public void onDownloadError(SpeedTestError speedTestError, String errorMessage) {
    System.out.println("Download error " + speedTestError + " : " + errorMessage);

    if ((speedTestError != SpeedTestError.FORCE_CLOSE_SOCKET) && (timer!=null)) {
            timer.purge();
            timer.cancel();
    }
}
```
It is still possible in theory to have 2 socket errors but I cant reproduce it anymore with the new version (1.09), but you are sure to have `SpeedTestError.FORCE_CLOSE_SOCKET` dispatched if this errors comes after  a ` speedTestSocket.forceStopTask()`

Let me know if you have any other issues

**Comment by freethinker6** [comment]
hi, that is working great.
I am getting an extra socket error besides from the FORCE_CLOSE_SOCKET when
I upload a file because I am stopping the task just like in the download
case. Is it possible you double check this please?

Thanks.

2016-05-13 17:45 GMT-05:00 Bertrand Martel <notifications@github.com>:

**Comment by akinaru** [comment]
Sorry for the waiting, this bug is fixed in v`1.10` : https://github.com/akinaru/speed-test-lib/commit/28c3d6b108604db2615e36e1c8f699f6004621fd

Thank you again for your feedback
